### PR TITLE
AHOYAPPS-638 Add Bandwidth Profile Settings

### PR DESCRIPTION
**Pull request by aaalaniz**
## Description

Add Bandwidth Profile Settings.

## Breakdown

- Add separate preference screen for configuring bandwidth profile settings. By default, the settings contain the recommended values for the Collaboration use case.
- Add BaseSettingsFragment to house boilerplate code between the different settings fragments.
- Update to latest BWP RC
- TODO @Alton09 changes

## Validation

- TODO @Alton09 changes

## Submission Checklist
 - [x] The source has been evaluated for semantic versioning changes and are reflected in ```versionName``` under `app/build.gradle`
 - [x] The `CHANGELOG.md` reflects any **feature**, **bug fixes**, or **known issues** made in the source code

### CODE REVIEW

File: app/src/main/java/com/twilio/video/app/data/Preferences.kt [hunk 1]
```diff
@@ -55,8 +58,37 @@ object Preferences {
...
+    const val BANDWIDTH_PROFILE_MAX_SUBSCRIPTION_BITRATE = "pref_bandwidth_profile_max_subscription_bitrate"
+    const val BANDWIDTH_PROFILE_MAX_SUBSCRIPTION_BITRATE_DEFAULT = 2400
+    const val BANDWIDTH_PROFILE_MAX_VIDEO_TRACKS = "pref_bandwidth_profile_max_video_tracks"
+    const val BANDWIDTH_PROFILE_MAX_VIDEO_TRACKS_DEFAULT = 5
+    const val BANDWIDTH_PROFILE_DOMINANT_SPEAKER_PRIORITY = "pref_bandwidth_profile_dominant_speaker_priority"
+    val BANDWIDTH_PROFILE_DOMINANT_SPEAKER_PRIORITY_DEFAULT = TrackPriority.STANDARD.name
+    const val BANDWIDTH_PROFILE_TRACK_SWITCH_OFF_MODE = "pref_bandwidth_profile_track_switch_off_mode"
+    const val BANDWIDTH_PROFILE_TRACK_SWITCH_OFF_MODE_DEFAULT = "Server Default"
+    const val BANDWIDTH_PROFILE_LOW_TRACK_PRIORITY_RENDER_DIMENSIONS = "pref_bandwidth_profile_low_track_priority_dimensions"
+    const val BANDWIDTH_PROFILE_LOW_TRACK_PRIORITY_RENDER_DIMENSIONS_DEFAULT = "Server Default"
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/data/Preferences.kt:74]: Add constant for "Server Default"

File: app/src/main/java/com/twilio/video/app/data/api/model/Topology.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (C) 2019 Twilio, Inc.
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/data/api/model/Topology.kt:2]:
```suggestion
 * Copyright (C) 2020 Twilio, Inc.
```

File: app/build.gradle [hunk 3]
```diff
@@ -50,6 +50,7 @@ android {
         multiDexEnabled true
 
         testInstrumentationRunner "com.twilio.video.app.InstrumentationRunner"
+        testInstrumentationRunnerArguments clearPackageData: 'true'
```
Comment (aaalaniz) [inline app/build.gradle:53]: Does this help with running tests locally?

File: app/src/androidTest/java/com/twilio/video/app/screen/BandwidthProfileSettingsScreen.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import com.twilio.video.app.util.getString
+import com.twilio.video.app.util.getStringArray
+
+fun clickBandwidthProfileSettings() {
+    val bandwidthProfile = R.string.settings_screen_bandwidth_profile
+    onView(withId(R.id.recycler_view))
+            .perform(actionOnItem<ViewHolder>(hasDescendant(withText(bandwidthProfile)), click()))
+}
+
+fun assertDefaultBandwidthProfileSettings() {
```
Comment (aaalaniz) [inline app/src/androidTest/java/com/twilio/video/app/screen/BandwidthProfileSettingsScreen.kt:29]: very nice 👍

File: app/src/main/java/com/twilio/video/app/ui/settings/BandwidthProfileSettingsFragment.kt [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            Preferences.BANDWIDTH_PROFILE_HIGH_TRACK_PRIORITY_RENDER_DIMENSIONS_DEFAULT)
+    }
+
+    override fun onResume() {
+        super.onResume()
+
+        (requireActivity() as BaseActivity).supportActionBar?.title = preferenceScreen.title
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/ui/settings/BandwidthProfileSettingsFragment.kt:45]: Can we move this to the BaseSettingsFragment?

File: app/src/main/java/com/twilio/video/app/ui/settings/BandwidthProfileSettingsFragment.kt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            Preferences.BANDWIDTH_PROFILE_STANDARD_TRACK_PRIORITY_RENDER_DIMENSIONS_DEFAULT)
+        setListPreferenceValue(R.array.settings_screen_bandwidth_profile_render_dimensions,
+            Preferences.BANDWIDTH_PROFILE_HIGH_TRACK_PRIORITY_RENDER_DIMENSIONS,
+            Preferences.BANDWIDTH_PROFILE_HIGH_TRACK_PRIORITY_RENDER_DIMENSIONS_DEFAULT)
+    }
+
+    override fun onResume() {
+        super.onResume()
+
+        (requireActivity() as BaseActivity).supportActionBar?.title = preferenceScreen.title
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/ui/settings/BandwidthProfileSettingsFragment.kt:42]: Can we move this to the BaseSettingsFragment?

File: app/src/main/java/com/twilio/video/app/ui/settings/SettingsFragment.kt [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+            // Return to login activity
+            loginIntent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
+            authenticator.logout()
+            startActivity(loginIntent)
+            activity.finishAffinity()
+        }
+    }
+
+    override fun onResume() {
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/ui/settings/SettingsFragment.kt:114]: Move to BaseSettingFragment

File: app/src/main/java/com/twilio/video/app/ui/settings/SettingsFragment.kt [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        super.onResume()
+
+        (requireActivity() as BaseActivity).supportActionBar?.title = preferenceScreen.title
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
```
Comment (aaalaniz) [inline app/src/main/java/com/twilio/video/app/ui/settings/SettingsFragment.kt:120]: Move to BaseSettingsFragment

File: app/build.gradle (hunk 3 above)
Comment (Alton09) [reply app/build.gradle:53]: Yeah it helps for local tests and tests on Firebase Test Lab. It basically clears all of the app cache and persisted data in between each test case. You no longer run into issues with login as each test starts in a logged out state.

**Pull request closed without merging**